mbedtls/src/ssl/async_io.rs

```diff
@@ -0,0 +1,253 @@
+            }
+        }
+
+        HandshakeFuture(self).await
+    }
+}
+
+impl<T: AsyncRead> AsyncRead for Context<T>
+where
+    for<'c, 'cx> (&'c mut TaskContext<'cx>, &'c mut T): IoCallbackUnsafe<AsyncStream>,
+{
+    fn poll_read(mut self: Pin<&mut Self>, cx: &mut TaskContext<'_>, buf: &mut ReadBuf<'_>) -> Poll<IoResult<()>> {
+        if self.handle().session.is_null() {
+            return Poll::Ready(Err(IoError::new(IoErrorKind::Other, "stream has been shutdown")));
+        }
+
+        self.with_bio_async(cx, |ssl_ctx| match ssl_ctx.recv(buf.initialize_unfilled()) {
+            Err(Error::SslPeerCloseNotify) => Poll::Ready(Ok(())),
+            Err(Error::SslWantRead) => Poll::Pending,
+            Err(e) => Poll::Ready(Err(crate::private::error_to_io_error(e))),
+            Ok(i) => {
+                buf.advance(i);
+                Poll::Ready(Ok(()))
+            }
+        })
+        .unwrap_or_else(|| Poll::Ready(Err(crate::private::error_to_io_error(Error::NetRecvFailed))))
+    }
+}
+
+impl<T: AsyncWrite + Unpin> AsyncWrite for Context<T>
+where
+    for<'c, 'cx> (&'c mut TaskContext<'cx>, &'c mut T): IoCallbackUnsafe<AsyncStream>,
+{
+    fn poll_write(mut self: Pin<&mut Self>, cx: &mut TaskContext<'_>, buf: &[u8]) -> Poll<IoResult<usize>> {
+        if self.handle().session.is_null() {
+            return Poll::Ready(Err(IoError::new(IoErrorKind::Other, "stream has been shutdown")));
+        }
+
+        let result = self
+            .with_bio_async(cx, |ssl_ctx| {
+                let buf = ssl_ctx.write_tracker.adjust_buf(buf)?;
+                match ssl_ctx.send(buf) {
+                    Err(Error::SslPeerCloseNotify) => Poll::Ready(Ok(0)),
+                    Err(Error::SslWantWrite) => Poll::Pending,
+                    Err(e) => Poll::Ready(Err(crate::private::error_to_io_error(e))),
+                    Ok(i) => Poll::Ready(Ok(i)),
+                }
+            })
+            .unwrap_or_else(|| Poll::Ready(Err(crate::private::error_to_io_error(Error::NetSendFailed))));
+
+        self.write_tracker.post_write(buf, &result);
+
+        cx.waker().clone().wake();
+
+        result
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut TaskContext<'_>) -> Poll<IoResult<()>> {
+        // We can only flush the actual IO here.
+        // To flush mbedtls we need writes with the same buffer until complete.
+        let io = &mut self
+            .io_mut()
+            .ok_or(IoError::new(IoErrorKind::Other, "stream has been shutdown"))?;
+        let stream = Pin::new(io);
+        stream.poll_flush(cx)
+    }
+
+    fn poll_shutdown(mut self: Pin<&mut Self>, cx: &mut TaskContext<'_>) -> Poll<IoResult<()>> {
+        if self.handle().session.is_null() {
+            return Poll::Ready(Err(IoError::new(IoErrorKind::Other, "stream has been shutdown")));
+        }
+
+        match self
+            .with_bio_async(cx, Context::close_notify)
```

Comment:
Are you allowed to call `close_notify` if `ssl_write` most recently returned `MBEDTLS_ERR_SSL_WANT_WRITE/READ` (i.e. if WriteTracker has pending data)?

Reply:
According to C code, it seems only `MBEDTLS_ERR_SSL_WANT_WRITE` affect `close_notify` behavior: it change to call and return `mbedtls_ssl_flush_output` instead.
So the above code should be okay :confused:
https://github.com/fortanix/rust-mbedtls/blob/fceaf689f879cc73c75393ad7b45ba238f7d6776/mbedtls-sys/vendor/library/ssl_msg.c#L5784-L5810